jerry-core/ecma/base/ecma-helpers-conversion.c

```diff
@@ -917,227 +919,512 @@ ecma_number_to_int32 (ecma_number_t num) /**< ecma-number */
+/**
+ * Implementation of the "grisu3" double to string
+ * conversion algorithm described in the research paper
+ *
+ * "Printing Floating-Point Numbers Quickly And Accurately with Integers"
+ * by Florian Loitsch, available at
+ * http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
+ *
+ * @return true - if success
+ *         false - otherwise
+ */
+static bool
+ecma_grisu3 (ecma_number_t num, /**< input ecma number */
+             char *buffer, /**< [out] digits buffer */
+             int *length, /**< [out] number of digits */
+             int *exp) /**< [out] exponent */
+{
+  int mk, kappa, success;
+  encoded_float_t dfp = ecma_number_to_encoded_fp (num);
+  encoded_float_t w = ecma_normalize_encoded_fp (dfp);
+
+  /* normalize boundaries */
+  encoded_float_t t = { (dfp.fract << 1) + 1, dfp.exp - 1 };
+  encoded_float_t b_plus = ecma_normalize_encoded_fp (t);
+  encoded_float_t b_minus;
+  encoded_float_t c_mk; /* Cached power of ten: 10^-k */
+  uint64_t *tmp_u64_p = (uint64_t *) &num;
+  uint64_t u64 = *tmp_u64_p;
+
+  /* Grisu only handles strictly positive finite numbers. */
+  JERRY_ASSERT (num > 0 && num <= 1.7976931348623157e308);
+
+  /* Is lower boundary is closer? */
+  if (!(u64 & GRISU3_FRACTION_MASK) && (u64 & GRISU3_EXPONENT_MASK) != 0)
   {
-    t /= 10;
-    digits_num++;
+    b_minus.fract = (dfp.fract << 2) - 1;
+    b_minus.exp =  dfp.exp - 2;
   }
+  else
+  {
+    b_minus.fract = (dfp.fract << 1) - 1;
+    b_minus.exp = dfp.exp - 1;
+  }
+
+  b_minus.fract = b_minus.fract << (b_minus.exp - b_plus.exp);
+  b_minus.exp = b_plus.exp;
+
+  mk = ecma_find_power_in_cache (GRISU3_MIN_TARGET_EXP - GRISU3_ENCODED_FLOAT_FRACT_SIZE - w.exp, &c_mk);
 
-  JERRY_ASSERT (digits_num > 0);
+  w = ecma_multiply_encoded_fp (w, c_mk);
+  b_minus = ecma_multiply_encoded_fp (b_minus, c_mk);
+  b_plus = ecma_multiply_encoded_fp (b_plus,  c_mk);
+
+  success = ecma_generate_digits (b_minus, w, b_plus, buffer, length, &kappa);
+  *exp = kappa - mk;
+
+  return success;
+} /* ecma_grisu3 */
+
+#endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
+
+/**
+  * Perform conversion of ecma-number to decimal representation with decimal exponent
+  *
+  * Note:
+  *      The calculated values correspond to s, n, k parameters in ECMA-262 v5, 9.8.1, item 5:
+  *         - s represents digits of the number;
+  *         - k is the number of digits;
+  *         - n is the decimal exponent.
+  */
+void
+ecma_number_to_decimal (ecma_number_t num, /**< ecma-number */
+                        uint64_t *out_digits_p, /**< [out] digits */
+                        int32_t *out_digits_num_p, /**< [out] number of digits */
+                        int32_t *out_decimal_exp_p) /**< [out] decimal exponent */
+{
+  JERRY_ASSERT (!ecma_number_is_nan (num));
+  JERRY_ASSERT (!ecma_number_is_zero (num));
+  JERRY_ASSERT (!ecma_number_is_infinity (num));
+
+#if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
+  char str[32];
```

Comment:
I know some followup work is needed after this patch, but this is good first step. Grisu3 has several benefits. First and the most important is that it fixes '3d-raytrace.js' of SunSpider. Grisu3 is more accurate then the current algorithm. Second, the patch decrease the code size by 4K. Third, although we use an additional uint64 to a string conversion it works with the same performance. After it is fixed in a followup patch, this could be even better. I think at the moment Grisu3 is definitely what we need. Especially for low end devices.